lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -252,256 +258,260 @@ impl Client {
+        }
+        let resp = if is_smart_card_op {
+            // Smart card control
+            let (code, res) = self.scard.ioctl(ioctl.io_control_code, payload)?;
+            if code == SPECIAL_NO_RESPONSE {
+                return Ok(vec![]);
+            }
+            DeviceControlResponse::new(&ioctl, code, res)
+        } else {
+            // Drive redirection, mimic FreeRDP's "no-op"
+            // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L677-L684
+            DeviceControlResponse::new(&ioctl, NTSTATUS::STATUS_SUCCESS.to_u32().unwrap(), vec![])
+        };
+        debug!("sending RDP: {:?}", resp);
+        let resp = self
+            .add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION, resp.encode()?)?;
+        Ok(resp)
+    }
+
+    fn process_irp_create(
+        &mut self,
+        device_io_request: DeviceIoRequest,
+        payload: &mut Payload,
+    ) -> RdpResult<Vec<Vec<u8>>> {
+        // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L207
+        let rdp_req = ServerCreateDriveRequest::decode(device_io_request, payload)?;
+        debug!("received RDP: {:?}", rdp_req);
+
+        // Send a TDP Shared Directory Info Request
+        // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L210
+        let tdp_req = SharedDirectoryInfoRequest::from(rdp_req.clone());
+        debug!("sending TDP: {:?}", tdp_req);
+        (self.tdp_sd_info_request)(tdp_req)?;
+
+        // Add a TDP Shared Directory Info Response handler to the handler cache.
+        // When we receive a TDP Shared Directory Info Response with this completion_id,
+        // this handler will be called.
+        self.pending_sd_info_resp_handlers.insert(
+            rdp_req.device_io_request.completion_id,
+            Box::new(
+                |cli: &mut Self, res: SharedDirectoryInfoResponse| -> RdpResult<Vec<Vec<u8>>> {
+                    let rdp_req = rdp_req;
+                    if res.err_code == TdpErrCode::Nil {
+                        // The file exists
+                        // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L214
+                        if res.fso.file_type == FileType::Directory {
+                            if rdp_req.create_disposition == flags::CreateDisposition::FILE_CREATE {
+                                // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L221
+                                // ERROR_ALREADY_EXISTS --> STATUS_OBJECT_NAME_COLLISION: https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L102
```

Comment:
What to these `x --> y` comments mean?

Reply:
The line above them points to the corollary code in FreeRDP where `x` is set (as a FreeRDP-type error), and the line itself points to the code in RDP where `x` is converted to `y`, the actual RDP code that's sent by FreeRDP. Its basically documenting how I determined which NTSTATUS to send back. Probably unnecessary to leave in, deleted.